### PR TITLE
azure function

### PR DESCRIPTION
**Pull request by jiminwen-msft**
Integrate azure function into sample SDK. So that hard-coding ACS endpoint, user_id and user token can be avoided.

### CODE REVIEW

File: samples/sample-chat-app/src/main/java/com/azure/android/communication/chat/sampleapp/UserTokenClient.java [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+                        + ". Error message: "
+                        + response.getBodyAsString()
+                    );
+                }
+
+                UserData userData = parseResponse(response.getBodyAsString());
+                userID = userData.getUserID();
+                userToken = userData.getUserToken();
+                ACSEndpoint = userData.getACSEndpoint();
+                latch.countDown();
```
Comment (vcolin7) [inline samples/sample-chat-app/src/main/java/com/azure/android/communication/chat/sampleapp/UserTokenClient.java:80]: I'm not sure if consuming the body twice could cause an error. We should do it only once.
```suggestion
                if (statusCode != 200) {
                    requestError[0] = new RuntimeException("Get token request failed with http status code "
                        + statusCode
                        + ". Error message: "
                        + response.getBodyAsString()
                    );
                } else {
                    UserData userData = parseResponse(response.getBodyAsString());
                    userId = userData.getUserId();
                    userToken = userData.getUserToken();
                    acsEndpoint = userData.getAcsEndpoint();
                    latch.countDown();
                }
```

File: samples/sample-chat-app/src/main/java/com/azure/android/communication/chat/sampleapp/MainActivity.java [hunk 2]
```diff
@@ -489,6 +489,7 @@ public void sendChatMessage(View view) {
                 logAndToast("Message sent with ID: " + chatMessageId);
             } catch (InterruptedException | ExecutionException e) {
                 logAndToast("Send message failed: " + e.getMessage());
+                logAndToast("Send message failed: " + e.getStackTrace());
```
Comment (vcolin7) [inline samples/sample-chat-app/src/main/java/com/azure/android/communication/chat/sampleapp/MainActivity.java:492]: We shouldn't repeat the same message at the beginning, people will think sending the message failed twice. Furthermore, toasting the stacktrace is not user friendly at all, we should just log it in my opinion.

File: samples/sample-chat-app/src/main/java/com/azure/android/communication/chat/sampleapp/UserTokenClient.java [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+ * Client that calls azure function to obtain user id and token
+ */
+public class UserTokenClient {
+    private final ClientLogger logger = new ClientLogger(UserTokenClient.class);
+
+    private final HttpClient httpClient;
+
+    private String endpoint;
+
+    private String userID;
```
Comment (vcolin7) [inline samples/sample-chat-app/src/main/java/com/azure/android/communication/chat/sampleapp/UserTokenClient.java:31]: nit;
```suggestion
    private String userId;
```

File: samples/sample-chat-app/src/main/java/com/azure/android/communication/chat/sampleapp/UserTokenClient.java [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+
+    private String acsEndpoint;
+
+    public UserTokenClient(String endpoint) {
+        httpClient = HttpClient.createDefault();
+        this.endpoint = endpoint;
+    }
+
+    public String getUserId() {
+        return userID;
```
Comment (vcolin7) [inline samples/sample-chat-app/src/main/java/com/azure/android/communication/chat/sampleapp/UserTokenClient.java:43]: nit;
```suggestion
        return userId;
```

File: samples/sample-chat-app/src/main/java/com/azure/android/communication/chat/sampleapp/UserTokenClient.java [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+    }
+
+    public String getUserToken() {
+        return userToken;
+    }
+
+    public String getACSEndpoint() {
```
Comment (vcolin7) [inline samples/sample-chat-app/src/main/java/com/azure/android/communication/chat/sampleapp/UserTokenClient.java:50]: nit;
```suggestion
    public String getAcsEndpoint() {
```

File: samples/sample-chat-app/src/main/java/com/azure/android/communication/chat/sampleapp/UserTokenClient.java [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+                        + statusCode
+                        + ". Error message: "
+                        + response.getBodyAsString()
+                    );
+                }
+                else {
+                    UserData userData = parseResponse(response.getBodyAsString());
+                    userID = userData.getUserID();
+                    userToken = userData.getUserToken();
+                    acsEndpoint = userData.getACSEndpoint();
```
Comment (vcolin7) [inline samples/sample-chat-app/src/main/java/com/azure/android/communication/chat/sampleapp/UserTokenClient.java:76]: nit;
```suggestion
                    userId = userData.getUserId();
                    userToken = userData.getUserToken();
                    acsEndpoint = userData.getAcsEndpoint();
```

File: samples/sample-chat-app/src/main/java/com/azure/android/communication/chat/sampleapp/UserTokenClient.java [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+        @JsonProperty("userId")
+        private String userId;
+
+        @JsonProperty("userToken")
+        private String userToken;
+
+        public UserData() {
+        }
+
+        public String getACSEndpoint() {
```
Comment (vcolin7) [inline samples/sample-chat-app/src/main/java/com/azure/android/communication/chat/sampleapp/UserTokenClient.java:127]: nit;
```suggestion
        public String getAcsEndpoint() {
```

File: samples/sample-chat-app/src/main/java/com/azure/android/communication/chat/sampleapp/UserTokenClient.java [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+            return acsEndpoint;
+        }
+
+        public String getUserID() {
```
Comment (vcolin7) [inline samples/sample-chat-app/src/main/java/com/azure/android/communication/chat/sampleapp/UserTokenClient.java:131]: nit;
```suggestion
        public String getUserId() {
```

**Review by vcolin7 — COMMENTED** [review]
Most things look good. Left a couple comments.

File: samples/sample-chat-app/src/main/java/com/azure/android/communication/chat/sampleapp/MainActivity.java (hunk 2 above)
Comment (jiminwen-msft) [reply samples/sample-chat-app/src/main/java/com/azure/android/communication/chat/sampleapp/MainActivity.java:492]: Removed this line. The toast operation shows a message on the application to notify a user something is wrong. This makes sense for a demo application from my opiontion

Comment (vcolin7) [reply samples/sample-chat-app/src/main/java/com/azure/android/communication/chat/sampleapp/MainActivity.java]: I understand the intent, I just think the stacktrace message might not be even readable in a toast. If you think it is, feel free to leave it in.

File: samples/sample-chat-app/src/main/java/com/azure/android/communication/chat/sampleapp/MyAppConfiguration.java [hunk 9]
```diff
@@ -29,6 +41,22 @@ public void accept(Throwable throwable) {
...
     public void onCreate() {
         super.onCreate();
+        try {
+            UserTokenClient userTokenClient = new UserTokenClient(AZURE_FUNCTION_URL);
+            //First user context
+            userTokenClient.getNewUserContext();
+            ACS_ENDPOINT = userTokenClient.getAcsEndpoint();
+            FIRST_USER_ID = userTokenClient.getUserId();
+            FIRST_USER_ACCESS_TOKEN = userTokenClient.getUserToken();
+            Log.i("debug", FIRST_USER_ID + FIRST_USER_ACCESS_TOKEN + ACS_ENDPOINT);
```
Comment (vcolin7) [inline samples/sample-chat-app/src/main/java/com/azure/android/communication/chat/sampleapp/MyAppConfiguration.java:51]:
```suggestion
            legger.info("debug", FIRST_USER_ID + FIRST_USER_ACCESS_TOKEN + ACS_ENDPOINT);
```

**Review by vcolin7 — APPROVED** [review]
Approving with a comment.

**Pull request merged**